### DISCUSSION TITLE
Automatically emitting rerun-if-changed instructions for cargo

### DISCUSSION THREAD

**Discussion by mickvangelderen** [hyperium/tonic#1019] [Q&A]
I find myself writing a build script similar to the following for every project:

```rust
fn main() {
    let protos = ["proto/rocsys/canonical.proto"];

    let includes = ["proto"];

    tonic_build::configure()
        .server_mod_attribute("rocsys.canonical", "#[cfg(feature = \"server\")]")
        .client_mod_attribute("rocsys.canonical", "#[cfg(feature = \"client\")]")
        .compile(&protos, &includes)
        .unwrap();

    for path in protos.iter().copied().chain(includes.iter().copied()) {
        println!("cargo:rerun-if-changed={}", path)
    }
}
```

I was wondering if we could let tonic emit those rerun-if-changed statements for us. It would be even better if we could narrow down the rerun-if-changed instructions only to the included files rather than asking cargo to monitor the entire include directories.

It is possible that this feature already exists but I couldn't find any signs of tonic_build writing to `stdout` nor does my build script print anything to `stdout` when removing my own `println` statements.

**Comment by davidpdrsn** [comment]
That feels like a bug to me. Either tonic-build or prost should do that. Wanna file an issue?

**Reply by mickvangelderen** [reply]
I can't say if it's a bug or a missing feature. I can imagine that the maintainers considered this to not be the responsibility of `tonic_build`. Filed an issue https://github.com/hyperium/tonic/issues/1020

**Reply by davidpdrsn** [reply]
Imo its a bug.
